internal/types/subtypes/registry.go

```diff
@@ -0,0 +1,81 @@
+package subtypes
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+type Subtype string
+
+const (
+	UnknownSubtype Subtype = "unknown"
+)
+
+func (t Subtype) String() string {
+	return string(t)
+}
+
+// Registry stores a collection of subtypes along with their prefixes and
+// allows for translating prefixes back to registered subtypes.
+type Registry struct {
+	subtypeMu        sync.RWMutex
+	subtypesPrefixes map[string]Subtype
+	knownSubtypes map[Subtype]interface{}
+}
+
+func NewRegistry() *Registry {
+	return &Registry{
+		subtypesPrefixes: make(map[string]Subtype),
+		knownSubtypes:    make(map[Subtype]interface{}),
+	}
+}
+
+// SubtypeFromType returns the Subtype from the provided string or if
+// no Subtype was registered with that string Unknown is returned.
+func (r *Registry) SubtypeFromType(t string) Subtype {
+	st := Subtype(t)
+	if _, ok := r.knownSubtypes[st]; !ok {
+		return UnknownSubtype
+	}
+	return st
+}
+
+// SubtypeFromId returns the Subtype from the provided id if the id's prefix
+// was registered with a Subtype. Otherwise Unknown is returned.
+func (r *Registry) SubtypeFromId(id string) Subtype {
+	i := strings.Index(id, "_")
+	if i == -1 {
+		return UnknownSubtype
+	}
+	prefix := id[:i]
+
+	r.subtypeMu.RLock()
+	subtype, ok := r.subtypesPrefixes[prefix]
+	r.subtypeMu.RUnlock()
+	if !ok {
+		return UnknownSubtype
+	}
+	return subtype
+}
+
+// Register registers all the prefixes for a provided Subtype. Register panics if the
```

Comment:
The locking in this file isn't fully consistent, and additionally panics are very much frowned upon at HC in application code. However -- is the intent (as I assume, from the rest of the code) that registration only happens at package init time? If so, I'd recommend documenting that as required behavior and just getting rid of the mutex. That would also make it "safer" to panic since it'd only happen at startup.